packages/lesswrong/components/ultraFeed/UltraFeedPostDialog.tsx

```diff
@@ -0,0 +1,287 @@
+
+const styles = defineStyles("UltraFeedPostDialog", (theme: ThemeType) => ({
+  dialogContent: {
+    padding: 20,
+    paddingTop: 0,
+    [theme.breakpoints.down('sm')]: {
+      padding: 10,
+      paddingTop: 0,
+    }
+  },
+  titleContainer: {
+    display: 'flex',
+    justifyContent: 'space-between',
+    alignItems: 'center',
+    paddingLeft: 16,
+    paddingRight: 16,
+    paddingTop: 8,
+    marginBottom: 8,
+  },
+  title: { 
+    fontFamily: theme.palette.fonts.sansSerifStack,
+    fontSize: '1.4rem',
+    fontWeight: 600,
+    opacity: 0.8,
+    lineHeight: 1.15,
+    textWrap: 'balance',
+    width: '100%',
+    '&:hover': {
+      opacity: 0.9,
+      textDecoration: 'none',
+    },
+    [theme.breakpoints.down('sm')]: {
+      fontSize: '1.6rem',
+    },
+  },
+  closeButton: {
+    fontFamily: theme.palette.fonts.sansSerifStack,
+    color: theme.palette.grey[500],
+    fontWeight: 600,
+    fontSize: "1.1rem",
+    cursor: 'pointer',
+    padding: 8,
+    '&:hover': {
+      color: theme.palette.grey[700],
+    },
+  },
+  dialogPaper: {
+    maxWidth: 765,
+    margin: 4
+  },
+  loadingContainer: {
+    display: "flex",
+    justifyContent: "center",
+    padding: "40px 0",
+  },
+  voteContainer: {
+    display: 'flex',
+    alignItems: 'center',
+    justifyContent: 'center',
+    padding: 16,
+    fontFamily: `${theme.palette.fonts.sansSerifStack} !important`,
+  },
+  voteBottom: {
+    position: 'relative',
+    fontSize: 42,
+    textAlign: 'center',
+    display: 'inline-block',
+    marginLeft: 'auto',
+    marginRight:'auto',
+    marginBottom: 40,
+    "@media print": { display: "none" },
+    '& h1': {
+      fontFamily: `${theme.palette.fonts.sansSerifStack} !important`,
+    }
+  },
+  scrolledHighlight: {
+    backgroundColor: `${theme.palette.primary.light}4c`,
+  },
+}));
+
+
+function escapeRegExp(s: string) {
+  return s.replace(/[-/\\^$*+?.()|[\]{}]/g, '\\$&');
+}
+
+/**
+ * Build a fuzzy regex that matches the snippet text allowing for 
+ * arbitrary whitespace and HTML tags between words
+ */
+function snippetRegex(snippet: string): RegExp {
+  // Collapse whitespace inside snippet first
+  const compact = snippet.trim().replace(/\s+/g, ' ');
+
+  // Allow ANY run of non‑letter/number characters (punctuation, whitespace) or HTML tags between words.
+  // This covers commas, periods, line‑breaks as <br>, etc.
+  const between = '(?:[^\\p{L}\\p{N}]+|<[^>]+>)+';
+
+  const pattern = compact
+    .split(' ')                  // individual words (already cleaned by generator)
+    .map(escapeRegExp)           // escape special regex chars (should be none but safe)
+    .join(between);
+
+  return new RegExp(pattern, 'iu');
+}
+
+/**
+ * Wraps the first occurrence of a snippet in the HTML with a span
+ * Handles whitespace differences between truncated and full HTML
+ */
+function wrapSnippet(html: string, snippet?: string, anchorId = 'snippet-anchor', highlightClass = '') {
```

Comment:
idk what exactly this is doing but we shouldn't be doing regex parsing (and then string concatenation!) of html that we're going to render unsafely

Reply:
(if this is for handling the "scroll to the section of the post that was truncated" thing, I think you should first try https://developer.mozilla.org/en-US/docs/Web/URI/Reference/Fragment/Text_fragments)